unicodetools/src/main/java/org/unicode/xml/UcdXML.java

```diff
@@ -0,0 +1,825 @@
+        }
+        return attributes;
+    }
+
+    private static boolean getIsAttributeIncluded(
+            String attrValue,
+            boolean isUnihanAttributeRange,
+            UcdPropertyDetail propDetail,
+            UcdProperty prop,
+            UCDXMLOUTPUTRANGE outputRange) {
+        if (attrValue == null) {
+            return false;
+        }
+        if (isUnihanAttributeRange) {
+            if (outputRange == UCDXMLOUTPUTRANGE.UNIHAN) {
+                if (prop.equals(UcdProperty.Numeric_Type) && !attrValue.equals("None")) {
+                    return true;
+                }
+                if (prop.equals(UcdProperty.Numeric_Value) && !attrValue.equals("NaN")) {
+                    return true;
+                }
+                return propDetail.isCJKAttribute()
+                        && (propDetail.isCJKShowIfEmpty() || !attrValue.isEmpty());
+            }
+            if (outputRange == UCDXMLOUTPUTRANGE.NOUNIHAN && propDetail.isCJKAttribute()) {
+                return false;
+            }
+            if (propDetail.isCJKShowIfEmpty()) {
+                return true;
+            }
+        }
+        if (propDetail.isBaseAttribute()) {
+            return true;
+        }
+        return !attrValue.isEmpty();
+    }
+
+    private static AttributesImpl getReservedAttributes(
+            VersionInfo version, AttributeResolver attributeResolver, ArrayList<Integer> range) {
+        AttributesImpl attributes = new AttributesImpl();
+
+        if (range.size() == 1) {
+            attributes.addAttribute(
+                    NAMESPACE, "cp", "cp", "CDATA", attributeResolver.getHexString(range.get(0)));
+        } else {
+            attributes.addAttribute(
+                    NAMESPACE,
+                    "first-cp",
+                    "first-cp",
+                    "CDATA",
+                    attributeResolver.getHexString(range.get(0)));
+            attributes.addAttribute(
+                    NAMESPACE,
+                    "last-cp",
+                    "last-cp",
+                    "CDATA",
+                    attributeResolver.getHexString(range.get(range.size() - 1)));
+        }
+        for (UcdPropertyDetail propDetail : UcdPropertyDetail.baseValues()) {
+            UcdProperty prop = propDetail.getUcdProperty();
+            if (version.compareTo(propDetail.getMinVersion()) >= 0
+                    && (propDetail.getMaxVersion() == null
+                            || version.compareTo(propDetail.getMaxVersion()) <= 0)) {
+                String attrValue =
+                        attributeResolver.getAttributeValue(
+                                propDetail.getUcdProperty(), range.get(0));
+
+                attributes.addAttribute(
+                        NAMESPACE, prop.getShortName(), prop.getShortName(), "CDATA", attrValue);
+            }
+        }
+        return attributes;
+    }
+
+    private static String getVersionString(VersionInfo version, int maxDigits) {
```

Comment:
I know that [VersionInfo.getVersionString()](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/VersionInfo.html#getVersionString-int-int-) is marked `@internal` but I suggest you call it anyway, rather than reinvent it.

Reply:
It's marked as `@Deprecated` not `@internal`. Is your recommendation still the same?

Reply:
yes

Reply:
Gone. Replaced with com.ibm.icu.util.VersionInfo#getVersionString